dev-tools/harness.js

```diff
@@ -0,0 +1,209 @@
+	constructor(options) {
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.format = options.format;
+		this.totalGames = options.totalGames;
+		this.all = !!options.all;
+		this.async = !!options.async;
+		this.p1options = options.p1options;
+		this.p2options = options.p2options;
+		// silence is golden (trumps noisy options)
+		this.silent = !!options.silent;
+		this.logs = !this.silent && !!options.logs;
+		this.verbose = !this.silent && !!options.verbose;
+		this.timer = options.timer;
+		this.formatter = options.formatter;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let timers = [];
+		let format, lastFormat;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				await Promise.all(games);
+				// TODO: trakkr still needs to support aggregated statistics...
+				// if (!this.silent) console.log(this.formatter.display(timers));
+				games = [];
+				timers = [];
+			}
+
+			const seed = this.prng.seed;
+			const timer = this.timer();
+
+			let battleStream;
+			try {
+				timer.start();
+				battleStream = new BattleStreams.BattleStream({timer});
+				const game = this.runGame(format, timer, battleStream).finally(() => timer.stop());
+				if (!this.async) {
+					await game;
+					if (this.verbose && timer.display) console.log(this.formatter.display(timer));
+				}
+				games.push(game);
+				timers.push(timer);
+			} catch (e) {
+				// TODO(#5297): danging promises still need to be fixed for this to work.
+				if (battleStream && battleStream.battle && this.logs) {
+					console.error(`${battleStream.battle.inputLog.join('\n')}\n\n`);
+				}
+				console.error(`Run \`node dev-tools/harness 1 --format=` +
+					`${format} --seed=${seed.join(',')} --verbose\` to debug:\n`, e);
+			}
+			lastFormat = format;
+		}
+
+		// Calculate how many games failed (failures weren't added to `games`).
+		return this.totalGames - (await Promise.all(games)).length;
+	}
+
+	async runGame(format, timer, battleStream) {
+		const t = timer.time('prepare');
+		const streams = BattleStreams.getPlayerStreams(battleStream || new BattleStreams.BattleStream());
+		// The seed used is the intial seed to begin with (its important that nothing
+		// advances the PRNG before the initial `runGame` call for repro purposes), but
+		// later is advanced by the four `newSeed()` calls, so each iteration should be
+		// 16 frames off the previous.
+		const spec = {formatid: format, seed: this.prng.seed};
+		const p1spec = {name: "Bot 1", seed: this.newSeed()};
+		const p2spec = {name: "Bot 2", seed: this.newSeed()};
+
+		const p1 = new RandomPlayerAI( // eslint-disable-line no-unused-vars
+			streams.p1, Object.assign({seed: this.newSeed()}, this.p1options));
+		const p2 = new RandomPlayerAI( // eslint-disable-line no-unused-vars
+			streams.p2, Object.assign({seed: this.newSeed()}, this.p2options));
+
```

Comment:
You forgot to ``start()`` the AIs?

Reply:
I didn't, this is before `start()` as added (they used to auto-start)